src/lib/circuit_value.ts

```diff
@@ -278,6 +288,64 @@ function circuitMain(
 }
 
 let primitives = new Set(['Field', 'Bool', 'Scalar', 'Group']);
+let complexTypes = new Set(['object', 'function']);
+
+// TODO properly type this at the interface
+// create recursive type that describes JSON-like structures of circuit types
+function circuitValue<T>(typeObj: any): AsFieldElements<T> {
```

Comment:
helper function which provides functionality equivalent to `CircuitValue`, but is less brittle, doesn't require decorators and can be applied in situations where a circuit value should be created from a plain JS object
i.e. 
```ts
let MyCircuitValue = circuitValue({ a: Group, b: Bool, c: [Field, Field] });
```
is equivalent to (but is more precisely typed than)
```ts
class MyCircuitValue extends CircuitValue {
  @prop a: Group;
  @prop b: Bool;
  @arrayProp(Field, 2) c: Field[];
  
  constructor(a: Group, b: Bool, c: Field[]) {
    this.a = a;
    this.b = b;
    this.c = c;
  }
}
```
`circuitValue` is only used in one place right now but I want to explore further usage, for example to simplify non-decorator ways of specifying circuit inputs

Reply:
Yeah I think we should move towards the more precise typed versions of things with generics in general.